[BFT] Node ejected before epoch recovery

durkmurder opened the pull request:
https://github.com/onflow/flow-go/issues/6331

### Context

This PR makes improvements on code clarity based on parent branch and implements a new test which ejects a node before submitting the recovery transaction which leads to a scenario where the DKG committee is a superset of consensus committee when recovery takes place. To support this behavior some changes to the tooling were made as well as vote processing logic to account for inequality of random beacon and consensus committees

jordanschalm commented on integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:
```diff
@@ -165,13 +175,103 @@ func (s *RecoverEpochSuite) TestRecoverEpoch() {
+	// start the paused collection node now that we are in EFM
+	require.NoError(s.T(), ln.Start())
+
+	// get final view form the latest snapshot
```
```suggestion
	// get final view from the latest snapshot
```

jordanschalm commented on integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:
```diff
@@ -165,13 +175,103 @@ func (s *RecoverEpochSuite) TestRecoverEpoch() {
+// 6. Ensure the network transitions into the recovery epoch and finalizes the first view of the recovery epoch.
+func (s *RecoverEpochSuite) TestRecoverEpochNodeEjected() {
+	// 1. Manually trigger EFM
+	// wait until the epoch setup phase to force network into EFM
+	s.AwaitEpochPhase(s.Ctx, 0, flow.EpochPhaseSetup, 10*time.Second, 500*time.Millisecond)
+
```
```suggestion
```
In principle, stopping the Collection Node _before_ entering the EpochSetup phase would be more reliable. The Consensus Nodes send off the first DKG message as soon as the EpochSetup phase starts, and can technically send their result submission after the DKG final view, so this could be race-prone.

jordanschalm commented on integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:
```diff
@@ -165,13 +175,103 @@ func (s *RecoverEpochSuite) TestRecoverEpoch() {
+	require.NoError(s.T(), err)
+
+	// Wait for at least the first view past the current epoch's original FinalView to be finalized.
+	// At this point we can observe that an extension has been added to the current epoch, indicating EFM.
```
```suggestion
```
We are using the epoch counter to observe EFM, not the presence of an extension.

jordanschalm commented on integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:
```diff
@@ -165,13 +175,103 @@ func (s *RecoverEpochSuite) TestRecoverEpoch() {
+
+	// assert transition to second epoch did not happen
+	// if counter is still 0, epoch emergency fallback was triggered as expected
+	s.AssertInEpoch(s.Ctx, 0)
```
I think this mechanism for checking EFM dates to before we added `EpochPhaseFallback`. We could retrieve the snapshot (currently line 248) a bit earlier, and assert that `snapshot.EpochPhase() == EpochPhaseFallback`, to make the test logic a little more explicit.

durkmurder commented on cmd/bootstrap/run/epochs.go:
```diff
@@ -55,9 +56,6 @@ func GenerateRecoverEpochTxArgs(log zerolog.Logger,
 
 	internalNodesMap := make(map[flow.Identifier]struct{})
 	for _, node := range internalNodes {
-		if !currentEpochIdentities.Exists(node.Identity()) {
```
[`d2448b9` (#6632)](https://github.com/onflow/flow-go/pull/6632/commits/d2448b9846556e038b053f2c46f806bd61bebd2d)

durkmurder commented on consensus/hotstuff/votecollector/combined_vote_processor_v2.go:
```diff
@@ -55,15 +55,20 @@ func (f *combinedVoteProcessorFactoryBaseV2) Create(log zerolog.Logger, block *m
+
+	publicKeyShares := make([]crypto.PublicKey, 0, dkg.Size())
+	for i := uint(0); i < dkg.Size(); i++ {
+		nodeID, err := dkg.NodeID(i)
```
Yeah, sounds good, I will add it.

durkmurder replied:
[`6e2a798` (#6632)](https://github.com/onflow/flow-go/pull/6632/commits/6e2a7987a2f54099bd6df32d84951b9f0a040ff0)

durkmurder commented on consensus/hotstuff/votecollector/combined_vote_processor_v3.go:
```diff
@@ -47,35 +47,43 @@ func (f *combinedVoteProcessorFactoryBaseV3) Create(log zerolog.Logger, block *m
+
 	// prepare the staking public keys of participants
 	stakingKeys := make([]crypto.PublicKey, 0, len(allParticipants))
+	stakingBeaconKeys := make([]crypto.PublicKey, 0, len(allParticipants))
```
I think this is a special case, it's basically beacon keys that take part in staking. There is a separate `beaconKeys` down the line that is used only for random beacon protocol.

durkmurder commented on integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:
```diff
@@ -165,13 +175,103 @@ func (s *RecoverEpochSuite) TestRecoverEpoch() {
+	require.Equal(s.T(), events[0].Events[0].Type, eventType)
+
+	// 5. Ensure the network transitions into the recovery epoch and finalizes the first view of the recovery epoch.
+	startViewOfNextEpoch := uint64(txArgs[1].(cadence.UInt64))
```
I haven't tried it but I can add it.

durkmurder commented:
During addressing comments found out that DKG is very inconsistent and flaky under this setup, left more details here: https://github.com/onflow/flow-go/issues/6729

AlexHentschel commented on integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:
I think there is one more test we could add. Lets denote 
* consensus committee 𝒞
* DKG committee 𝒟

To my understanding, test `TestRecoverEpoch` verifies recovery for 𝒞 =  𝒟.
Test `TestRecoverEpochNodeEjected` checks the scenario  𝒞 ∪ {ρ} =  𝒟 for some node ρ.

It would be great to also verify that scenario  𝒞 =  𝒟∪ {ρ} for some node ρ works as expected. Essentially, this could be the case if node ρ successfully staked for epoch N+1, but then the DKG for epoch N+1 failed and we recovered the network but still decided to include ρ, as it already locked up its stake.

durkmurder replied:
I have a more general case to cover: https://github.com/onflow/flow-go/issues/6645. How about I take care of this in separate issue by implementing this test?

AlexHentschel reviewed (approved):
Very nice work.

durkmurder commented on consensus/hotstuff/votecollector/combined_vote_processor_v3.go:
```diff
@@ -47,39 +47,33 @@ func (f *combinedVoteProcessorFactoryBaseV3) Create(log zerolog.Logger, block *m
-	}
-
-	rbSigAggtor, err := signature.NewWeightedSignatureAggregator(allParticipants, beaconKeys, msg, msig.RandomBeaconTag)
+	beaconAggregator, err := signature.NewWeightedSignatureAggregator(allParticipants, beaconKeys, msg, msig.RandomBeaconTag)
```
This is true, this wasn't properly tested as we don't use combined vote processor v3 in integration tests. I will add a unit test for this case.

durkmurder replied:
[`3a94957` (#6632)](https://github.com/onflow/flow-go/pull/6632/commits/3a94957ebdafcc5633d3fd6a36d1c96c75ec2f94)
